analyze-network-with-subnetwork-trace/src/main/java/com/esri/arcgismaps/sample/analyzenetworkwithsubnetworktrace/MainActivity.kt

```diff
@@ -0,0 +1,512 @@
+import com.google.android.material.snackbar.Snackbar
+import com.google.android.material.textfield.TextInputEditText
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+
+class MainActivity : AppCompatActivity() {
+
+    private val TAG = MainActivity::class.java.simpleName
+
+    // set up data binding for the activity
+    private val activityMainBinding: ActivityMainBinding by lazy {
+        DataBindingUtil.setContentView(this, R.layout.activity_main)
+    }
+
+    private val sourceDropdown: AutoCompleteTextView by lazy {
+        activityMainBinding.sourceDropdown
+    }
+
+    private val operatorDropdown: AutoCompleteTextView by lazy {
+        activityMainBinding.operatorDropdown
+    }
+
+    private val expressionTextView: TextView by lazy {
+        activityMainBinding.expressionTextView
+    }
+
+    private val valuesDropdown: AutoCompleteTextView by lazy {
+        activityMainBinding.valuesDropdown
+    }
+
+    private val valuesBackgroundView: RelativeLayout by lazy {
+        activityMainBinding.valuesBackgroundView
+    }
+
+    private val valueBooleanButton: ToggleButton by lazy {
+        activityMainBinding.valueBooleanButton
+    }
+
+    private val valuesEditText: TextInputEditText by lazy {
+        activityMainBinding.valuesEditText
+    }
+
+    private val barriersCheckbox: CheckBox by lazy {
+        activityMainBinding.barriersCheckBox
+    }
+
+    private val containersCheckbox: CheckBox by lazy {
+        activityMainBinding.containersCheckbox
+    }
+
+    private val utilityNetwork by lazy {
+        UtilityNetwork(getString(R.string.utility_network_url))
+    }
+
+    private var initialExpression: UtilityTraceConditionalExpression? = null
+    private var sourceTier: UtilityTier? = null
+    private var utilityTraceConfiguration: UtilityTraceConfiguration? = null
+    private var sourcesList: List<UtilityNetworkAttribute>? = null
+    private var operatorsList: Array<UtilityAttributeComparisonOperator>? = null
+    private var startingLocation: UtilityElement? = null
+    private var codedValuesList: List<CodedValue>? = null
+    private var sourcePosition: Int = 0
+    private var operatorPosition: Int = 0
+    private var valuePosition: Int = 0
+    private var dialog: AlertDialog? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        ArcGISEnvironment.applicationContext = this
+        ArcGISEnvironment.authenticationManager.arcGISAuthenticationChallengeHandler =
+            getAuthenticationChallengeHandler()
+
+        // create and display the loading dialog
+        showLoadingDialog(true)
+
+        // load the utility network
+        lifecycleScope.launch {
+            utilityNetwork.load().getOrElse {
+                return@launch showError("Error loading utility network: " + it.message.toString())
```

Comment:
Before this line, `showLoadingDialog(false)` should be called, otherwise, it will be showing there forever, if an error occurs.